doc/neps/nep-0023-backwards-compatibility.rst

```diff
@@ -0,0 +1,290 @@
+errors or having a better user experience for new users.  These competing
+needs often give rise to heated debates and delays in accepting or rejecting
+contributions.  This NEP tries to address that by providing a policy as well
+as examples and rationales for when it is or isn't a good idea to break
+backwards compatibility.
+
+General principles:
+
+- Aim not to break users' code unnecessarily.
+- Aim never to change code in ways that can result in users silently getting
+  incorrect results from their previously working code.
+- Backwards incompatible changes can be made, provided the benefits outweigh
+  the costs.
+- When assessing the costs, keep in mind that most users do not read the mailing
+  list, do not look at deprecation warnings, and sometimes wait more than one or
+  two years before upgrading from their old version.  And that NumPy has
+  many hundreds of thousands or even a couple of million users, so "no one will
+  do or use this" is very likely incorrect.
+- Benefits include improved functionality, usability and performance (in order
+  of importance), as well as lower maintenance cost and improved future
+  extensibility.
+- Bug fixes are exempt from the backwards compatibility policy.  However in case
+  of serious impact on users (e.g. a downstream library doesn't build anymore),
+  even bug fixes may have to be delayed for one or more releases.
+- The Python API and the C API will be treated in the same way.
+
+
+Examples
+^^^^^^^^
+
+We now discuss a number of concrete examples to illustrate typical issues
+and trade-offs.
+
+**Changing the behavior of a function**
+
+``np.histogram`` is probably the most infamous example.
+First, a new keyword ``new=False`` was introduced, this was then switched
+over to None one release later, and finally it was removed again.
+Also, it has a ``normed`` keyword that had behavior that could be considered
+either suboptimal or broken (depending on ones opinion on the statistics).
+A new keyword ``density`` was introduced to replace it; ``normed`` started giving
+``DeprecationWarning`` only in v.1.15.0.  Evolution of ``histogram``::
+
+    def histogram(a, bins=10, range=None, normed=False):  # v1.0.0
+
+    def histogram(a, bins=10, range=None, normed=False, weights=None, new=False):  #v1.1.0
+
+    def histogram(a, bins=10, range=None, normed=False, weights=None, new=None):  #v1.2.0
+
+    def histogram(a, bins=10, range=None, normed=False, weights=None):  #v1.5.0
+
+    def histogram(a, bins=10, range=None, normed=False, weights=None, density=None):  #v1.6.0
+
+    def histogram(a, bins=10, range=None, normed=None, weights=None, density=None):  #v1.15.0
+        # v1.15.0 was the first release where `normed` started emitting
+        # DeprecationWarnings
+
+The ``new`` keyword was planned from the start to be temporary; such a plan
+forces users to change their code more than once.  Such keywords (there have
+been other instances proposed, e.g. ``legacy_index`` in
```

Comment:
I don't think `legacy_index` falls in the same category as `new` here. It is not intended as temporary -- there is no plan to remove it.

Reply:
> there is no plan to remove it.

there isn't such a plan right now:) I'm pretty sure the next generation of developers will look at it the same way as we look at things that are ugly and we'd rather not have - we start proposing to deprecate and remove.

Reply:
True, although if a typical downstream project supports say 4 years of numpy versions, I think once those have passed and people start discussing such a plan, then maybe it is a bit of a different thing than a temporary `new` kwarg.

Reply:
@shoyer removed. clearly your proposal wasn't in the same league as `new` which disappeared in 2 releases. 

I'm serious though that everything named `new`, `old`, `legacy`, `change_behavior` etc. is a mistake almost by definition. It was similar in the `random` NEP; thinking about a separate use case (unit testing) and creating `StableRandom` was a much better outcome there than the original `LegacyGenerator`.